### PR TITLE
feat: Replace default Go user-agent with oauth2-proxy and version

### PR DESCRIPTION
**Pull request by middagj**
Changing the default user agent from `Go-http-client/1.1` to `oauth2-proxy/x.y.z`.

## Description

I moved the `VERSION` to `pkg` and use it to replace the default user agent from the Go standard library with one that specifies `oauth2-proxy` with the current version. This is done by deriving a httpClient from the default one and add a `Transport` that sets the user-agent if not already set. Then using this httpClient in the requests package and setting it on the context for the oauth2 and oidc libraries.

## Motivation and Context

We have an authentication service that is behind Akamai which blocks generic user agents, like `Go-http-client/1.1`.
Using a more specific one fixes this. Also it can help to detect faulty clients when running an IdP related to oauth2-proxy or specific versions. This is a reason that Let's Encrypt have a strong policy to have people use an actual user agent instead of the standard library's one.

I thought about making the default user agent also configurable, but that is a bit more work if globals are not an option.
Also tried first to add default user-agent via the http builder, but then found out that if done also in the oauth2 and oidc library you need to customize it via the httpClient.

## How Has This Been Tested?

There is already a test that test this specific case, so I left it.
I did roll out a patched version on our environment and it fixed out issues.
If needed, I could add some more tests for the oidc case.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.
- [x] I have written tests for my code changes.

### CODE REVIEW

File: providers/oidc.go [hunk 1]
```diff
@@ -83,6 +85,7 @@ func (p *OIDCProvider) Redeem(ctx context.Context, redirectURL, code, codeVerifi
 		},
 		RedirectURL: redirectURL,
 	}
+	ctx = oidc.ClientContext(ctx, requests.DefaultHttpClient)
```
Comment (tuunit) [inline providers/oidc.go:88]: This seems like a lot of duplication. Would it be possible to configure it globally for all outgoing requests?

File: pkg/requests/http.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package requests
+
+import (
+	"net/http"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg"
+)
+
+type userAgentTransport struct {
+}
```
Comment (tuunit) [inline pkg/requests/http.go:10]: using an empty struct seems off to me

File: providers/oidc.go (hunk 1 above)
Comment (tuunit) [reply providers/oidc.go:88]: Maybe have a look at
https://github.com/go-swagger/go-swagger/issues/1563#issuecomment-393223595

**Review by tuunit — CHANGES_REQUESTED** [review]

File: providers/oidc.go (hunk 1 above)
Comment (middagj) [reply providers/oidc.go:88]: Yes, that is a more sophisticated way, I can change it to.

> This seems like a lot of duplication. Would it be possible to configure it globally for all outgoing requests?

No, it would require changes to the oauth2 and oidc library

File: pkg/requests/http.go (hunk 2 above)
Comment (middagj) [reply pkg/requests/http.go]: Resolved by using previous suggestion.

File: providers/oidc.go (hunk 1 above)
Comment (middagj) [reply providers/oidc.go]: Changed in b6da03604981fec57adbaf37e7e7d16e6cddd6a8

File: pkg/requests/http.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+func (t *userAgentTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	r2 := req.Clone(req.Context())
+	setDefaultUserAgent(r2.Header, t.userAgent)
+	return t.next.RoundTrip(r2)
+}
+
+var DefaultHTTPClient = &http.Client{Transport: &userAgentTransport{
+	http.DefaultTransport,
+	"oauth2-proxy/" + pkg.VERSION,
```
Comment (JoelSpeed) [inline pkg/requests/http.go:22]: Use named parameters not positional please

File: pkg/version.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package main
+package pkg
```
Comment (JoelSpeed) [inline pkg/version.go:1]: Move to a dedicated version package under `pkg/version` please

File: Makefile [hunk 5]
```diff
@@ -40,7 +40,7 @@ lint: validate-go-version
 build: validate-go-version clean $(BINARY)
 
 $(BINARY):
-	CGO_ENABLED=0 $(GO) build -a -installsuffix cgo -ldflags="-X main.VERSION=${VERSION}" -o $@ github.com/oauth2-proxy/oauth2-proxy/v7
+	CGO_ENABLED=0 $(GO) build -a -installsuffix cgo -ldflags="-X github.com/oauth2-proxy/oauth2-proxy/v7/pkg.VERSION=${VERSION}" -o $@ github.com/oauth2-proxy/oauth2-proxy/v7
```
Comment (JoelSpeed) [inline Makefile:43]: This needs to be updated as well

Comment (middagj) [reply Makefile:43]: Yes, forgot about that one.

**Comment by JoelSpeed** [comment]
Something's not right here, when I look at the files view, or directly on your branch, it is showing as pkg.Version in the Makefile still, can you double check?

**Comment by middagj** [comment]
> Something's not right here, when I look at the files view, or directly on your branch, it is showing as pkg.Version in the Makefile still, can you double check?

Strange.  If I look locally, there are no `pkg.VERSION` anymore. Also on GitHub UI not: https://github.com/oauth2-proxy/oauth2-proxy/pull/2570/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52

**Comment by JoelSpeed** [comment]
UI still showing pkg.Version in the makefile for me, tried different browsers and the GitHub app, perhaps rebase and force push?

**Comment by middagj** [comment]
> UI still showing pkg.Version in the makefile for me, tried different browsers and the GitHub app, perhaps rebase and force push?

Done.

**Comment by JoelSpeed** [comment]
It has changed now, but shouldn't it be pkg/version.VERSION?

File: Makefile [hunk 6]
```diff
@@ -40,7 +40,7 @@ lint: validate-go-version
 build: validate-go-version clean $(BINARY)
 
 $(BINARY):
-	CGO_ENABLED=0 $(GO) build -a -installsuffix cgo -ldflags="-X main.VERSION=${VERSION}" -o $@ github.com/oauth2-proxy/oauth2-proxy/v7
+	CGO_ENABLED=0 $(GO) build -a -installsuffix cgo -ldflags="-X github.com/oauth2-proxy/oauth2-proxy/v7/pkg.version.VERSION=${VERSION}" -o $@ github.com/oauth2-proxy/oauth2-proxy/v7
```
Comment (tuunit) [inline Makefile:43]: github.com/oauth2-proxy/oauth2-proxy/v7/pkg.version.VERSION

Is the . between pkg and version correct here?

In the import statements you use "github.com/oauth2-proxy/oauth2-proxy/v7/pkg/version"

Comment (tuunit) [reply Makefile:43]: I'll test it on my end tomorrow

**Comment by middagj** [comment]
> It has changed now, but shouldn't it be pkg/version.VERSION?

No, it should be import path:

> -X importpath.name=value
> Set the value of the string variable in importpath named name to value.
> This is only effective if the variable is declared in the source code either uninitialized

https://pkg.go.dev/cmd/link

I also tested this.

**Comment by JoelSpeed** [comment]
The import path _is_ `pkg/version` not `pkg.version` no?

**Comment by middagj** [comment]
> The import path _is_ `pkg/version` not `pkg.version` no?

Ah yes, you are right. I thought you only meant the package name. I will fix it.

File: Makefile [hunk 7]
```diff
@@ -40,7 +40,7 @@ lint: validate-go-version
 build: validate-go-version clean $(BINARY)
 
 $(BINARY):
-	CGO_ENABLED=0 $(GO) build -a -installsuffix cgo -ldflags="-X main.VERSION=${VERSION}" -o $@ github.com/oauth2-proxy/oauth2-proxy/v7
+	CGO_ENABLED=0 $(GO) build -a -installsuffix cgo -ldflags="-X github.com/oauth2-proxy/oauth2-proxy/v7/pkg/version.VERSION=${VERSION}" -o $@ github.com/oauth2-proxy/oauth2-proxy/v7
```
Comment (tuunit) [inline Makefile:43]: Please apply the same change in `dist.sh` as it is used as part of `make release`

File: pkg/requests/builder.go [hunk 8]
```diff
@@ -58,7 +58,7 @@ func (r *builder) WithMethod(method string) Builder {
 
 // WithHeaders replaces the request header map with the given header map.
 func (r *builder) WithHeaders(header http.Header) Builder {
-	r.header = header
+	r.header = header.Clone()
```
Comment (tuunit) [inline pkg/requests/builder.go:61]: Why is this change necessary?

**Review by tuunit — CHANGES_REQUESTED** [review]

File: pkg/requests/builder.go (hunk 8 above)
Comment (middagj) [reply pkg/requests/builder.go:61]: Ah, this is a remnant of an earlier attempt to add it in the `Builder` interface. In the current code base I don't think it is strictly necessary, but if you combine `WithHeaders` and `SetHeader` you are mutating the original header you passed. I left it because it look strange to me, but I can change it if that is preferred.

File: pkg/requests/http.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+type userAgentTransport struct {
+	next      http.RoundTripper
+	userAgent string
+}
+
+func (t *userAgentTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	r2 := req.Clone(req.Context())
+	setDefaultUserAgent(r2.Header, t.userAgent)
+	return t.next.RoundTrip(r2)
```
Comment (tuunit) [inline pkg/requests/http.go:17]:
```suggestion
	r := req.Clone(req.Context())
	setDefaultUserAgent(r.Header, t.userAgent)
	return t.next.RoundTrip(r)
```

File: providers/oidc.go [hunk 10]
```diff
@@ -83,6 +85,7 @@ func (p *OIDCProvider) Redeem(ctx context.Context, redirectURL, code, codeVerifi
 		},
 		RedirectURL: redirectURL,
 	}
+	ctx = oidc.ClientContext(ctx, requests.DefaultHTTPClient)
```
Comment (tuunit) [inline providers/oidc.go:88]:
```suggestion
	}

	ctx = oidc.ClientContext(ctx, requests.DefaultHTTPClient)
```

**Review by tuunit — CHANGES_REQUESTED** [review]

**Comment by middagj** [comment]
Still waiting for @JoelSpeed and @tuunit

**Review by JoelSpeed — APPROVED** [review]
